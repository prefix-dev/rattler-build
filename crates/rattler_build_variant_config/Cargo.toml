[package]
name = "rattler_build_variant_config"
version = "0.1.0"
authors.workspace = true
edition.workspace = true
license.workspace = true
readme = "README.md"
repository.workspace = true
description = "Variant configuration system for rattler-build"

[features]
default = ["parser"]
parser = ["minijinja", "miette"]

[dependencies]
# Core serde
serde = { workspace = true }
serde_yaml = { workspace = true }

# Rattler crates
rattler_conda_types = { workspace = true }

# Rattler-build crates
rattler_build_types = { workspace = true }
rattler_build_jinja = { workspace = true }
rattler_build_yaml_parser = { path = "../rattler_build_yaml_parser", version = "0.1.0" }

# YAML parsing
marked-yaml = { version = "0.8.0" }

# Template rendering (optional with 'parser' feature)
minijinja = { version = "2.11.0", features = [
  "unstable_machinery",
  "custom_syntax",
], optional = true }

# Error handling
thiserror = "2.0.17"
miette = { workspace = true, optional = true }

# Utilities
tracing = { workspace = true }
fs-err.workspace = true

[dev-dependencies]
insta = { workspace = true, features = ["yaml", "filters"] }
