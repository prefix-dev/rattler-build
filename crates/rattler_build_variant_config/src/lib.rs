//! # rattler_build_variant_config
//!
//! A standalone library for managing variant configurations in conda package builds.
//!
//! This crate provides functionality for:
//! - Loading variant configurations from YAML files (`variants.yaml`)
//! - Loading legacy `conda_build_config.yaml` files with selector support
//! - Computing all possible variant combinations (build matrices)
//! - Handling "zip keys" to synchronize related variants
//!
//! ## Example
//!
//! ```rust
//! use rattler_build_types::NormalizedKey;
//! use rattler_build_variant_config::VariantConfig;
//! use std::collections::HashSet;
//!
//! # fn main() -> Result<(), Box<dyn std::error::Error>> {
//! // Parse a variant configuration
//! let yaml = r#"
//! python:
//!   - "3.9"
//!   - "3.10"
//! numpy:
//!   - "1.20"
//!   - "1.21"
//! zip_keys:
//!   - [python, numpy]
//! "#;
//!
//! let config = VariantConfig::from_yaml_str(yaml)?;
//!
//! // Define which variables are actually used
//! let mut used_vars = HashSet::new();
//! used_vars.insert("python".into());
//! used_vars.insert("numpy".into());
//!
//! // Compute all combinations
//! let combinations = config.combinations(&used_vars, None)?;
//!
//! // With zip_keys, we get 2 combinations (not 3x2=6)
//! assert_eq!(combinations.len(), 2);
//! # Ok(())
//! # }
//! ```
//!
//! ## Variant Configuration Format
//!
//! A variant configuration is a YAML file that defines variables and their possible values:
//!
//! ```yaml
//! # Simple variants - creates a full cartesian product
//! python:
//!   - "3.9"
//!   - "3.10"
//! compiler:
//!   - gcc
//!   - clang
//!
//! # This creates 2x2 = 4 combinations:
//! # [python=3.9, compiler=gcc]
//! # [python=3.9, compiler=clang]
//! # [python=3.10, compiler=gcc]
//! # [python=3.10, compiler=clang]
//! ```
//!
//! ## Zip Keys
//!
//! Zip keys allow you to synchronize related variants:
//!
//! ```yaml
//! python:
//!   - "3.9"
//!   - "3.10"
//! numpy:
//!   - "1.20"
//!   - "1.21"
//! zip_keys:
//!   - [python, numpy]
//!
//! # This creates only 2 combinations:
//! # [python=3.9, numpy=1.20]
//! # [python=3.10, numpy=1.21]
//! ```
//!
//! ## Advanced Features (with `parser` feature)
//!
//! When the `parser` feature is enabled (default), you can use conditionals and Jinja expressions:
//!
//! ```yaml
//! python:
//!   - if: unix
//!     then: ["3.14", "3.15"]
//!   - if: win
//!     then: ["3.14"]
//!
//! foobar:
//!   - ${{ "unknown" if unix else "known" }}
//! ```
//!
//! ## conda_build_config Support
//!
//! This crate also supports loading legacy `conda_build_config.yaml` files:
//!
//! ```yaml
//! python:
//!   - 3.9
//!   - 3.10  # [unix]
//!   - 3.11  # [osx]
//! ```

pub mod combination;
pub mod conda_build_config;
pub mod config;
pub mod error;

#[cfg(feature = "parser")]
pub mod parser;

// Re-export main types
pub use combination::compute_combinations;
pub use conda_build_config::{SelectorContext, load_conda_build_config};
pub use config::VariantConfig;
pub use error::{VariantConfigError, VariantError, VariantExpandError};

#[cfg(feature = "parser")]
pub use parser::{parse_variant_file, parse_variant_str};

#[cfg(test)]
mod tests {
    use super::*;
    use std::collections::HashSet;

    #[test]
    fn test_basic_workflow() {
        let yaml = r#"
python:
  - "3.9"
  - "3.10"
numpy:
  - "1.20"
  - "1.21"
"#;

        let config = VariantConfig::from_yaml_str(yaml).unwrap();

        let mut used_vars = HashSet::new();
        used_vars.insert("python".into());
        used_vars.insert("numpy".into());

        let combos = config.combinations(&used_vars, None).unwrap();
        assert_eq!(combos.len(), 4); // 2x2 combinations
    }

    #[test]
    fn test_with_zip_keys() {
        let yaml = r#"
python:
  - "3.9"
  - "3.10"
numpy:
  - "1.20"
  - "1.21"
zip_keys:
  - [python, numpy]
"#;

        let config = VariantConfig::from_yaml_str(yaml).unwrap();

        let mut used_vars = HashSet::new();
        used_vars.insert("python".into());
        used_vars.insert("numpy".into());

        let combos = config.combinations(&used_vars, None).unwrap();
        assert_eq!(combos.len(), 2); // Zipped: only 2 combinations
    }
}
