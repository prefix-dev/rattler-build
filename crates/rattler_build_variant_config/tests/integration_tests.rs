use rattler_build_variant_config::{SelectorContext, VariantConfig, load_conda_build_config};
use rattler_conda_types::Platform;
use std::collections::HashSet;
use std::path::PathBuf;

fn test_data_dir() -> PathBuf {
    PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("test-data")
}

#[test]
fn test_simple_variant_config() {
    let path = test_data_dir().join("simple/variants.yaml");
    let config = VariantConfig::from_file(&path).unwrap();

    insta::assert_yaml_snapshot!("simple_config", config);
}

#[test]
fn test_simple_combinations() {
    let path = test_data_dir().join("simple/variants.yaml");
    let config = VariantConfig::from_file(&path).unwrap();

    // Test with just python and numpy
    let mut used_vars = HashSet::new();
    used_vars.insert("python".into());
    used_vars.insert("numpy".into());

    let combinations = config.combinations(&used_vars, None).unwrap();
    insta::assert_yaml_snapshot!("simple_python_numpy_combos", combinations);
}

#[test]
fn test_simple_all_combinations() {
    let path = test_data_dir().join("simple/variants.yaml");
    let config = VariantConfig::from_file(&path).unwrap();

    // Test with all variables
    let mut used_vars = HashSet::new();
    used_vars.insert("python".into());
    used_vars.insert("numpy".into());
    used_vars.insert("compiler".into());

    let combinations = config.combinations(&used_vars, None).unwrap();
    insta::assert_yaml_snapshot!("simple_all_combos", combinations);

    // Should be 3 python × 3 numpy × 2 compiler = 18 combinations
    assert_eq!(combinations.len(), 18);
}

#[test]
fn test_zip_keys_config() {
    let path = test_data_dir().join("with_zip_keys/variants.yaml");
    let config = VariantConfig::from_file(&path).unwrap();

    insta::assert_yaml_snapshot!("zip_keys_config", config);
}

#[test]
fn test_zip_keys_combinations() {
    let path = test_data_dir().join("with_zip_keys/variants.yaml");
    let config = VariantConfig::from_file(&path).unwrap();

    // Test with python and numpy (should be zipped)
    let mut used_vars = HashSet::new();
    used_vars.insert("python".into());
    used_vars.insert("numpy".into());

    let combinations = config.combinations(&used_vars, None).unwrap();
    insta::assert_yaml_snapshot!("zip_keys_python_numpy_combos", combinations);

    // Should be 3 combinations (zipped), not 9
    assert_eq!(combinations.len(), 3);
}

#[test]
fn test_zip_keys_all_combinations() {
    let path = test_data_dir().join("with_zip_keys/variants.yaml");
    let config = VariantConfig::from_file(&path).unwrap();

    // Test with all variables (two zip groups)
    let mut used_vars = HashSet::new();
    used_vars.insert("python".into());
    used_vars.insert("numpy".into());
    used_vars.insert("compiler".into());
    used_vars.insert("compiler_version".into());

    let combinations = config.combinations(&used_vars, None).unwrap();
    insta::assert_yaml_snapshot!("zip_keys_all_combos", combinations);

    // Should be 3 (python/numpy zipped) × 2 (compiler/version zipped) = 6
    assert_eq!(combinations.len(), 6);
}

#[test]
fn test_conda_build_config_linux() {
    let path = test_data_dir().join("conda_build_config/conda_build_config.yaml");
    let context = SelectorContext::new(Platform::Linux64);

    let config = load_conda_build_config(&path, &context).unwrap();
    insta::assert_yaml_snapshot!("conda_build_config_linux", config);
}

#[test]
fn test_conda_build_config_osx() {
    let path = test_data_dir().join("conda_build_config/conda_build_config.yaml");
    let context = SelectorContext::new(Platform::OsxArm64);

    let config = load_conda_build_config(&path, &context).unwrap();
    insta::assert_yaml_snapshot!("conda_build_config_osx", config);
}

#[test]
fn test_conda_build_config_win() {
    let path = test_data_dir().join("conda_build_config/conda_build_config.yaml");
    let context = SelectorContext::new(Platform::Win64);

    let config = load_conda_build_config(&path, &context).unwrap();
    insta::assert_yaml_snapshot!("conda_build_config_win", config);
}

#[test]
fn test_multi_file_merge() {
    let base = test_data_dir().join("multi_file/base.yaml");
    let override_file = test_data_dir().join("multi_file/override.yaml");

    let config = VariantConfig::from_files(&[base, override_file]).unwrap();
    insta::assert_yaml_snapshot!("multi_file_merged", config);

    // Python should be overridden to ["3.11", "3.12"]
    let python_variants = config.get(&"python".into()).unwrap();
    assert_eq!(python_variants.len(), 2);
    assert_eq!(python_variants[0].to_string(), "3.11");
}

#[test]
fn test_multi_file_combinations() {
    let base = test_data_dir().join("multi_file/base.yaml");
    let override_file = test_data_dir().join("multi_file/override.yaml");

    let config = VariantConfig::from_files(&[base, override_file]).unwrap();

    let mut used_vars = HashSet::new();
    used_vars.insert("python".into());
    used_vars.insert("numpy".into());
    used_vars.insert("compiler".into());

    let combinations = config.combinations(&used_vars, None).unwrap();
    insta::assert_yaml_snapshot!("multi_file_combos", combinations);

    // Should be 2 python/numpy (zipped) × 1 compiler = 2 combinations
    assert_eq!(combinations.len(), 2);
}

#[test]
fn test_filtered_combinations() {
    let path = test_data_dir().join("simple/variants.yaml");
    let config = VariantConfig::from_file(&path).unwrap();

    let mut used_vars = HashSet::new();
    used_vars.insert("python".into());
    used_vars.insert("compiler".into());

    // Filter to only python=3.10
    let mut filter = std::collections::BTreeMap::new();
    filter.insert("python".into(), "3.10".into());

    let combinations = config.combinations(&used_vars, Some(&filter)).unwrap();
    insta::assert_yaml_snapshot!("filtered_python_310_combos", combinations);

    // Should only have 2 combinations (python=3.10 with gcc and clang)
    assert_eq!(combinations.len(), 2);

    // All should have python=3.10
    for combo in &combinations {
        assert_eq!(combo.get(&"python".into()).unwrap().to_string(), "3.10");
    }
}

#[test]
fn test_partial_variable_usage() {
    let path = test_data_dir().join("simple/variants.yaml");
    let config = VariantConfig::from_file(&path).unwrap();

    // Only use python (ignore numpy and compiler)
    let mut used_vars = HashSet::new();
    used_vars.insert("python".into());

    let combinations = config.combinations(&used_vars, None).unwrap();
    insta::assert_yaml_snapshot!("partial_python_only_combos", combinations);

    // Should be 3 combinations (just python variants)
    assert_eq!(combinations.len(), 3);
}
