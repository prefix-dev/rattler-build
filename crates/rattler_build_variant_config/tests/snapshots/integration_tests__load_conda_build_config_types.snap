---
source: crates/rattler_build_variant_config/tests/integration_tests.rs
expression: config
---
zip_keys:
  - - c_compiler_version
    - cxx_compiler_version
    - fortran_compiler_version
    - docker_image
  - - cudnn
    - cuda_compiler_version
  - - python
    - numpy
    - python_impl
  - - arrow_cpp
    - libarrow
abseil_cpp:
  - "20220623.0"
alsa_lib:
  - 1.2.8
arb:
  - "2.23"
arpack:
  - "3.7"
arrow_cpp:
  - 11.0.0
  - 10.0.1
  - 9.0.0
  - 8.0.1
assimp:
  - 5.2.5
aws_sdk_cpp:
  - "4.5"
blas_impl:
  - openblas
boolean:
  - true
integer:
  - "5"
libblas:
  - 3.9 *netlib
libcblas:
  - 3.9 *netlib
liblapack:
  - 3.9 *netlib
liblapacke:
  - 3.9 *netlib
noboolean:
  - "true"
nointeger:
  - "5"
